include/upipe/uprobe.h

```diff
@@ -300,20 +300,24 @@ static inline int uprobe_throw_next(struct uprobe *uprobe, struct upipe *upipe,
     return uprobe_throw_va(uprobe->next, upipe, event, args);
 }
 
-/** @internal @This throws a log event. This event is thrown whenever a pipe
- * wants to send a textual message.
+/** @internal @This throws a log event, with vprintf-style message generation.
  *
  * @param uprobe pointer to probe hierarchy
  * @param upipe description structure of the pipe
  * @param level level of importance of the message
- * @param msg textual message
+ * @param format format string of the textual message
+ * @param args arguments for the format string
  */
-static inline void uprobe_log(struct uprobe *uprobe, struct upipe *upipe,
-                              enum uprobe_log_level level, const char *msg)
+static inline void uprobe_vlog(struct uprobe *uprobe, struct upipe *upipe,
+                               enum uprobe_log_level level,
+                               const char *format,
+                               va_list args)
 {
     struct ulog ulog;
-    ulog_init(&ulog, level, msg);
+    ulog_init(&ulog, level, format);
```

Comment:
Maybe you should remove `ulog_init` as it doesn't initialise `ulog.args`?

Reply:
`ulog.args` is initialized now.